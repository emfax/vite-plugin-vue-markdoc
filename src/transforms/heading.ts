import { Tag } from "@markdoc/markdoc";
import type { Node, Schema } from "@markdoc/markdoc";
import slugify from "slugify";

// Or replace this with your own function
function generateID(node: Node, attributes: { [key: string]: unknown }) {
  if (attributes.id && typeof attributes.id === 'string') {
    return attributes.id;
  }

  let id = "";

  for(const child of node.walk()) {
    if (child.type === "inline") continue;

    if (child.type === "text") {
      const text = child.attributes.content as string;

      id = slugify(text);
    }
  }

  return id;
}

export const heading = {
  children: ["inline"],
  
  attributes: {
    id: { type: String },
    level: { type: Number, required: true, default: 1 },
  },
  
  transform(node: Node, config): Tag {
    const children = node.transformChildren(config);

    const id = generateID(node, node.attributes);

    return new Tag(
      `h${node.attributes['level']}`,
      { ...(id && { id }) },
      children,
    );
  }
} as Schema;
