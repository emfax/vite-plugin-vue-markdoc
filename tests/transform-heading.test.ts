import { readFile } from "node:fs/promises";
import { beforeAll, describe, expect, test } from "vitest";
import Markdoc from "@markdoc/markdoc"; 
import heading from "../src/transforms/heading";

let md = "";

describe('heading transformer', () => {
  beforeAll(async () => {
    md = await readFile("tests/fixtures/heading.md", "utf-8");
  });

  test("the correct id is generated", () => {
    const ast = Markdoc.parse(md);
    const renderable = Markdoc.transform(ast, { nodes: { heading }});
    const html = Markdoc.renderers.html(renderable);

    expect(html).toMatchSnapshot();
  });
});
