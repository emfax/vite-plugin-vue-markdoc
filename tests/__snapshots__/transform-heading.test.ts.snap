// Vitest Snapshot v1

exports[`heading transformer > the correct id is generated 1`] = `"<article><h1 id=\\"introduction\\">Introduction</h1></article>"`;
